policies/recipes/steps.go

```diff
@@ -90,8 +100,335 @@ func parsePermissions(s string) (os.FileMode, error) {
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(decompressed, reader)
+		decompressed.Close()
+		if err != nil {
+			return err
+		}
+		return extractTar(decompressed.Name(), step.ArchiveExtraction.Destination)
+	default:
+		return fmt.Errorf("Unrecognized archive type %q", step.ArchiveExtraction.GetType())
+	}
+}
+
+func zipIsDir(name string) bool {
+	return strings.HasSuffix(name, string(os.PathSeparator))
+}
+
+func extractZip(zipPath string, dst string) error {
+	zr, err := zip.OpenReader(zipPath)
+	if err != nil {
+		return err
+	}
+	defer zr.Close()
+
+	// Check for conflicts
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, f.FileHeader.Name))
+		if err != nil {
+			return err
+		}
+		fmt.Printf("checking if file %s exists\n", filen)
+		stat, err := os.Stat(filen)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if zipIsDir(filen) && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
+
+	// Create dirs
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, f.FileHeader.Name))
+		if err != nil {
+			return err
+		}
+
+		if !zipIsDir(filen) {
+			continue
+		}
+		_, err = os.Stat(filen)
+		if err == nil {
+			continue
+		}
+		if !os.IsNotExist(err) {
+			return err
+		}
+		err = os.MkdirAll(filen, 0755)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create files.
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, f.Name))
+		if err != nil {
+			return err
+		}
+		if zipIsDir(filen) {
+			continue
+		}
+		filedir := filepath.Dir(filen)
+		_, err = os.Stat(filedir)
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(filedir, 0755)
+		}
+		if err != nil {
+			return err
+		}
+		fmt.Printf("os.Create %s\n", filen)
+		reader, err := f.Open()
+		if err != nil {
+			return err
+		}
+
+		dst, err := os.OpenFile(filen, os.O_RDWR|os.O_CREATE, 0755)
+		if err == nil {
+			_, err = io.Copy(dst, reader)
+		}
+		reader.Close()
+
+		if err != nil {
+			return err
+		}
+		err = os.Chtimes(filen, time.Now(), f.Modified)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func extractTar(tarName string, dst string) error {
+	file, err := os.Open(tarName)
+	if err != nil {
+		return err
+	}
+	tr := tar.NewReader(file)
+
+	// Check for conflicts
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		fmt.Printf("checking if file %s exists\n", filen)
+		stat, err := os.Stat(filen)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if header.Typeflag == tar.TypeDir && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
+
+	file.Seek(0, 0)
```

Comment:
btw this was just the testing mechanism i was using, i haven't determined if there's a better way to rewind a tar

Reply:
Okay, I didn't find a way around reading the file multiple times and the tar reader doesn't seem to provide a way to go backward at all so I'll keep this implementation for now.